packages/ra-ui-materialui/src/list/Datagrid.tsx

```diff
@@ -99,12 +99,14 @@ const Datagrid: FC<DatagridProps> = props => {
     const updateSort = useCallback(
         event => {
             event.stopPropagation();
-            setSort(
-                event.currentTarget.dataset.sort,
-                event.currentTarget.dataset.order
-            );
+            const newField = event.currentTarget.dataset.field;
+            const newOrder =
+                currentSort.field === newField && currentSort.order === 'ASC'
+                    ? 'DESC'
+                    : 'ASC';
```

Comment:
The default sort order can de defined on a per field basis. You must rely on the `data-order` attribute.

Reply:
In this new piece of code, if the user clicks on the header of a column that isn't yet used for sorting (thus `currentSort.field !== newField`), the sort order defaults to the `ASC` sort order. In the previous piece of code, the sort order defaulted to `event.currentTarget.dataset.order`. That's a breaking change. 

You see, developers can customize the default sort order for a field using the `sortByOrder` prop on a Field - and that prop is stored in the header dataset. If you don't use that value, you're breaking the `sortByOrder` feature. 

See https://marmelab.com/react-admin/List.html#specify-sort-order